node/src/components/fetcher/trie_fetcher.rs

```diff
@@ -76,44 +79,74 @@ where
             .collect();
         bytesrepr::deserialize(data)
     }
+
+    fn next_peer(&mut self, prev_peer: I) -> Option<&I> {
+        // return None early if not found - it means we didn't even request the chunk from the
+        // previous peer
+        let index = self
+            .peers
+            .iter()
+            .enumerate()
+            .rev() // to optimize slightly for the happy path: we expect prev_peer to be the last
+            .find(|(_index, peer)| **peer == prev_peer)
+            .map(|(index, _)| index)?;
+        self.peers.remove(index);
+        self.peers.last()
+    }
+
+    fn merge(&mut self, other: PartialChunks<I>) {
+        self.chunks.extend(other.chunks);
+        self.responders.extend(other.responders);
+        self.peers.extend(other.peers);
```

Comment:
Ah, definitely. I don't know how I missed that :man_facepalming: 
A set will be a good idea. I was also thinking of making this an intersection of the two sets instead of the sum - as if a peer isn't in one of the sets, it probably means that we already tried downloading from them and failed. i'm not sure if it's a good idea, though, maybe it's better to err on the side of retrying too much rather than too little?

Reply:
On second thought, a set might not be such a good idea, actually - since this also acts as a queue of peers to try, and we want it randomized deterministically, I'd expect. A `BTreeSet` would have them sorted, and a `HashSet` would shuffle them indeterministically. With a `Vec` we can shuffle them once and store the order, so I think that's the best option...